types/decimal.go

```diff
@@ -0,0 +1,324 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"strconv"
+	"strings"
+	"testing"
+)
+
+// NOTE: never use new(Dec) or else we will panic unmarshalling into the
+// nil embedded big.Int
+type Dec struct {
+	*big.Int `json:"int"`
+}
+
+// number of decimal places
+const Precision = 10
+
+func precisionInt() *big.Int {
+	return new(big.Int).Exp(big.NewInt(10), big.NewInt(Precision), nil)
+}
+
+// nolint - common values
+func ZeroDec() Dec { return Dec{big.NewInt(0)} }
+func OneDec() Dec  { return Dec{precisionInt()} }
+
+// get the precision multiplier
+func precisionMultiplier(prec int64) *big.Int {
+	if prec > Precision {
+		panic("too much precision")
```

Comment:
If we validate all created decimals, is there a reason for this check?

Reply:
it's a sanity check, hence a panic not an error return. Going to leave in here for now (we can remove or disable sanity for optimization later in the future if we want)